### PR TITLE
Structify VARIANT and use in more places

### PR DESCRIPTION
**Pull request by hughbe**
Extracted from #3197 

This PR is limited to making `VARIANT` blittable, using it in more places (and fixing some memory bugs in x64!) and fixing a bug made in porting `~VARENUM.TYPEMASK` instead of `VARENUM.TYPEMASK`

The idea is that this should be semi-trivial to merge and will make the above PR easier to review as functionality in that PR should be limited to making `ToObject` actually work in the way it should

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3326)

### CODE REVIEW

File: src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANT.cs [hunk 1]
```diff
@@ -23,32 +27,30 @@ public sealed class VARIANT
...
+            public void Dispose() => Clear();
 
             public object ToObject()
             {
                 IntPtr val = data1;
                 long longVal;
 
-                VARENUM vtType = vt & ~VARENUM.TYPEMASK;
+                VARENUM vtType = vt & VARENUM.TYPEMASK;
                 switch (vtType)
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANT.cs:54]:
```suggestion
                switch (Type)
```

**Comment by RussKie** [comment]
LGTM but I'd like to get @JeremyKuhne and @weltkante to weigh in here.

**Comment by hughbe** [comment]
I have fixed the leak - could you just double check that is what you meant?

File: src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANT.cs [hunk 2]
```diff
@@ -166,8 +167,10 @@ public object ToObject()
                         return (val != IntPtr.Zero);
 
                     case VARENUM.VARIANT:
-                        VARIANT varStruct = Marshal.PtrToStructure<VARIANT>(val);
+                    {
+                        using VARIANT varStruct = Marshal.PtrToStructure<VARIANT>(val);
                         return varStruct.ToObject();
+                    }
```
Comment (weltkante) [inline src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANT.cs:173]: Sorry, lost track a bit of what the other PR did. Wasn't this previously just returning the VARIANT directly? Now I'm confused, I remember a discussion turning it recursive.

If it already is recursive I _think_ disposing is wrong, PtrToStructure will just see native pointers and not copy anything, so there is nothing to dispose. To be clean about the fact that you don't have to dispose you could cast to pointer and call ToObject, alternatively adding a comment why not to dispose.

**Comment by hughbe** [comment]
I extracted the test changes from this PR into https://github.com/dotnet/winforms/pull/3349#event-3372402657

Hopefully that should reduce the size of this PR

File: src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANT.cs (hunk 2 above)
Comment (hughbe) [reply src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANT.cs]: Went with
```
                        VARIANT* pVariant = (VARIANT*)data;
                        return pVariant->ToObject();
```

File: src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANT.cs [hunk 3]
```diff
@@ -23,33 +27,30 @@ public sealed class VARIANT
...
+                fixed (VARIANT* pThis = &this)
                 {
-                    Oleaut32.SysFreeString(data1);
+                    VariantClear(pThis);
                 }
 
-                data1 = data2 = IntPtr.Zero;
                 vt = VARENUM.EMPTY;
+                data1 = IntPtr.Zero;
+                data1 = IntPtr.Zero;
```
Comment (weltkante) [inline src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANT.cs:43]: nit: redundant assignment
```suggestion
                data1 = IntPtr.Zero;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs [hunk 4]
```diff
@@ -559,45 +559,36 @@ public void DetachEventHandler(string eventName, EventHandler eventHandler)
...
-            public object var;
-            public byte b;
-        }
-        private static readonly int VariantSize = (int)Marshal.OffsetOf(typeof(FindSizeOfVariant), "b");
-
         /// <summary>
         ///  Convert a object[] into an array of VARIANT, allocated with CoTask allocators.
         /// </summary>
-        internal unsafe static IntPtr ArrayToVARIANTVector(object[] args)
+        internal unsafe static Oleaut32.VARIANT* ArrayToVARIANTVector(object[] args)
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs:565]: It would be preferable to just pin an allocated array of `VARIANT`. Ultimately we want to remove native Alloc* calls wherever we can to avoid accidental native leaks and the overhead of trying to prevent said leaks. I'd suggest creating a ref struct with a Dispose() to encapsulate creation and disposal.

We can iterate on this to push the array to the stack for small sets in the future if we really need to, but I would suspect that we won't. We could also potentially rent from the array pool, but I also don't think we should start there.

``` C#
internal ref struct VARIANTVector
{
    public VARIANT[] Variants;
    public VARIANTVector(object[] values) { /* Create array */ } 
    public void Dispose() { /* Free contents of array */ }
}

// Elsewhere

using var vector = new VARIANTVector(values);
fixed (VARIANT* v = vector.Variants)
{
}
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2IManagedPerPropertyBrowsingHandler.cs [hunk 5]
```diff
@@ -258,47 +258,36 @@ private static string[] GetStringsFromPtr(IntPtr ptr, uint cStrings)
...
 
-        private static object[] GetVariantsFromPtr(IntPtr ptr, uint cVariants)
+        private unsafe static object[] GetVariantsFromPtr(Oleaut32.VARIANT* ptr, uint cVariants)
         {
-            if (ptr != IntPtr.Zero)
+            if (ptr == null)
             {
-                object[] objects = new object[cVariants];
-                IntPtr curVariant;
+                return new object[cVariants];
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2IManagedPerPropertyBrowsingHandler.cs:265]: This is bizarre and hard to follow (not your fault, obviously). It would be far better to return null and handle it in the caller.

File: src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs (hunk 4 above)
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs:565]: Will experiment with this after #3349 is merged - could you let me know how to action that?

Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs:565]: Sorry, I was OOF last week. I'm trying to get through PRs as quickly as I can. :)

File: src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2IManagedPerPropertyBrowsingHandler.cs (hunk 5 above)
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2IManagedPerPropertyBrowsingHandler.cs]: Done

File: src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs (hunk 4 above)
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs]: Cleaned this up, obviously will verify when the above PR is in :)

File: src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANTVector.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            public VARIANT[] Variants;
+
+            public VARIANTVector(object[] values)
+            {
+                if (values == null)
+                {
+                    Variants = Array.Empty<VARIANT>();
+                }
+
+                var variants = new VARIANT[values.Length];
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANTVector.cs:23]: This will throw NRE, we must `return` at line:21

File: src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs [hunk 7]
```diff
@@ -428,49 +426,49 @@ public unsafe object InvokeScript(string scriptName, object[] args)
...
-                        dispParams.rgdispidNamedArgs = IntPtr.Zero;
-                        dispParams.cNamedArgs = 0;
+                        return null;
+                    }
 
-                        object[] retVals = new object[1];
-                        var pExcepInfo = new Oleaut32.EXCEPINFO();
+                    if (args != null)
+                    {
+                        // Reverse the arg order so that parms read naturally after IDispatch. (
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs:436]: [nit] Looks like a stray `(` at EoL.

File: src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANTVector.cs (hunk 6 above)
Comment (hughbe) [reply src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.VARIANTVector.cs:24]: fixed, tests should verify this (had to rebase to pull them in)

File: src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs (hunk 7 above)
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/HtmlDocument.cs]: odd. generally cleaned up this
